### ISSUE TITLE
Plans to include Python convenience API for shape analysis?

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#1040]
Is there a plan to include convenience functions (i.e. to avoid Python OCC) for common tasks when a shape is created? e.g. volume / area calcs, check intersection, centroid?

**Comment by aothms** [comment]
I think most of this exists:

check intersection: https://github.com/IfcOpenShell/IfcOpenShell/pull/130 (uses the OCCT in C++ IfcOpenShell)

volume / area calcs: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcgeom/IfcGeomRepresentation.h#L80 `calc_area()` etc. not sure if it's accessible from Python also uses OCCT in C++ IfcOpenShell

Can you create a complete overview of desired functionality?

**Comment by Moult** [comment]
Fantastic! I had no idea that existed! I'll record them in the OSArch Wiki to help expose it to users and will definitely make use of them!

I was having issues with fcl in IFC clash reporting clashes on objects that merely have faces that are coplanar, and was wondering if functions existed in IfcOpenShell that would make my life easier. Is there a function that deals with mesh-mesh intersection with consideration of possible co-planarness?

I was also planning to start writing MicroMVDs for geometric requirements, and so those functions would be useful. I didn't really have a list of desired functionality, but wanted to explore what was already created :)

**Comment by Moult** [comment]
Are there checks for manifold / closed geometry? This would be useful to verify whether we can later on recalculate volumes for costing / quantity take-off.

**Comment by Moult** [comment]
Ping @fbpyr you were searching for this I believe.

**Comment by aothms** [comment]
> Are there functions for checking for manifold / closed geometry? This would be useful to verify whether we can later on recalculate volumes for costing / quantity take-off.

Sorry, I missed that, calculate_volume() does actually a check like that, but it's not very informative. Also it can give partial results when there are multiple representation items. How can we improve that?

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcgeom/IfcGeomRepresentation.cpp#L217

**Comment by Moult** [comment]
Just bumping this, since it recently cropped up again in [conversation](https://community.osarch.org/discussion/comment/12249#Comment_12249) it would be nice to have:

 - derive_volume()
 - derive_area()
 - derive_centroid()
 - derive_bbox()

**Comment by Moult** [comment]
Quoting from @aothms in the discussion since it seems discussions don't get referenced: https://github.com/IfcOpenShell/IfcOpenShell/discussions/2514#discussioncomment-3912555

@aothms are there already functions for centroid/area/volume?

**Comment by Moult** [comment]
I've noticed the iterator has functions like compute bounds and bound_min and bound_max, but couldn't figure out how to actually run those functions (e.g. `compute_bounds(True)`, and then `bound_min()` gives a swig gp_XYZ but I couldn't access the individual X/Y/Z.

**Comment by VolgardTheScientist** [comment]
Hi Moult, I am trying to compute bounding box dimensions for all my elements in an ifc file. I tried the get_bbox but it often results with false dims when items are rotated from the global X/Y axes. I tried this with both get_vertices and get_element_vertices but with no success. From this post I understand that there already is a way to get this to work. Is there any chance you (or any other helpful soul) could please point me in the right direction, or ideally a code snippet? Many thanks!

**Comment by Moult** [comment]
If you've tried with both get_vertices (i.e. object local coords) and get_element_vertices (i.e. absolute coordinates) with no success, then unfortunately it sounds like your data doesn't have proper object aligned axes.

If you send me a sample file / object I can help confirm.

**Comment by VolgardTheScientist** [comment]
Hi Moult, many thanks for the speedy reply! I have attached two sample files, one of a totaly stripped down room [(source -> ArchiCAD )](https://drive.google.com/file/d/1hrxTGxM6HuhpFGOhze6EKMZ2-HmOlvsP/view?usp=sharing), rotated from XY global axes -> take a look at the slab. I am getting some very strange behaviour from get_bbox -> e.g. IfcDoors and IfcWalls turn out fine, but the IfcSlab and IfcCovering don't. You can see similar problem with the other test file -> a small section of a structural engineer's [model from Revit](https://drive.google.com/file/d/1knMabtLeEWUXUOG8xVO80hLxMxTth42d/view?usp=drive_link). I am getting the same behaviour with all my test models, where objects don't align with X/Y axis.
I have raised this also on OSarch, you will see my code [here](https://community.osarch.org/discussion/1685/boundingbox-issues/p1?new=1). Many thanks for your help!

**Comment by Moult** [comment]
Yeah indeed this is just a problem with how ArchiCAD works. See the screenshots below which demonstrate how walls and doors have their placement axes aligning (correctly) with their geometry whereas slabs follow the global orientation and ignore the slab profile.

![2023-09-16-200007_709x765_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/e08d5361-b9d8-4931-a79c-9ee1f919931a)
![2023-09-16-200001_752x749_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/9acaa05c-ac89-46c1-a0b0-616f9119df9e)
![2023-09-16-195954_743x723_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/cc7963d3-9b7c-4460-b6fe-12fa5746ef20)

In short, since your IFC is missing the critical data for axis orientation, you will need to calculate the Oriented Bounding Box OBB yourself. This will show you how in Blender: https://blender.stackexchange.com/questions/261049/how-to-get-an-objects-oriented-bounding-box-obb-using-script

**Comment by VolgardTheScientist** [comment]
Thanks for the speedy reply - very helpful! Too bad that both ArchiCAD and Revit have the same issue. The OBB in Blender works great, but the bpy package works only inside of Blender - I need to get this to work in a online environment with pure ifcopenshell. 
**Get_OBB** would be a great addition to the ifcopenshell.

**Comment by Moult** [comment]
@VolgardTheScientist the bpy module is only used to get the active object. After that it's pure numpy. It would be a great contribution to add get_obb to the util.shape module. Would you be interested in writing a PR? It should be pretty straightforward to use that code.

**Comment by VolgardTheScientist** [comment]
I have no other choice than to try ;) My app won't work without it anyway. Let's see if I can get this to work with my limited skills!

**Comment by Moult** [comment]
Good luck! You should be able to copy paste almost all of the code and only make a few minor modifications. We're here if you need help!
